src/AH/Hardware/ExtendedInputOutput/ShiftRegisterIn.hpp

```diff
@@ -0,0 +1,241 @@
+     *          output to a shift register is not useful.
+     */
+    void analogWrite(pin_t, analog_t) override // LCOV_EXCL_LINE
+        __attribute__((deprecated)) {}         // LCOV_EXCL_LINE
+
+    /**
+     * @copydoc analogWrite
+     */
+    void analogWriteBuffered(pin_t, analog_t) override // LCOV_EXCL_LINE
+        __attribute__((deprecated)) {}                 // LCOV_EXCL_LINE
+
+    /**
+     * @brief   Initialize the shift register: set the pin mode of the address pins
+     *          and the enable pin to output mode.
+     */
+    void begin() override;
+
+    void read();
+
+    /**
+     * @brief   No periodic updating of the state is necessary, all actions are 
+     *          carried out when the user calls analogRead or digitalRead.
+     */
+    void updateBufferedOutputs() override {} // LCOV_EXCL_LINE
+
+    /**
+     * @brief   Reads the current values of the inputs connected to the shift
+     *          register and stores them for retrieval via digitalRead.
+     * 
+     */
+    void updateBufferedInputs() override;
+
+  private:
+    const pin_t dataPin;
+    const pin_t clockPin;
+    const pin_t clockEnablePin;
+    const pin_t loadPin; 
+
+    uint8_t buffer;
+
+    /**
+     * @brief   Enable the shift register.
+     */
+    void prepareReading();
+
+    /**
+     * @brief   Disable the shift register.
+     */
+    void afterReading();
+};
+
+/**
+ * @brief   An alias for ShiftRegisterIn<8> to use with CD74HC165 shift 
+ *          registers.
+ * 
+ * @ingroup AH_ExtIO
+ */
+using CD74HC165 = ShiftRegisterIn<8>;
+
+// -------------------------------------------------------------------------- //
+
+template <uint8_t N>
+void ShiftRegisterIn<N>::pinMode(pin_t, PinMode_t mode) {
+    ExtIO::pinMode(dataPin, mode);
+}
+
+template <uint8_t N>
+void ShiftRegisterIn<N>::pinModeBuffered(pin_t, PinMode_t mode) {
+    ShiftRegisterIn<N>::pinMode(dataPin, mode);
+}
+
+template <uint8_t N>
+int ShiftRegisterIn<N>::digitalRead(pin_t pin) {
+    return bitRead(buffer, pin);
+}
+
+template <uint8_t N>
+int ShiftRegisterIn<N>::digitalReadBuffered(pin_t pin) {
+    // ISSUE: bitRead is available when building for Arduino but not when 
+    // running tests
+    return bitRead(buffer, pin);
+}
+
+template <uint8_t N>
+analog_t ShiftRegisterIn<N>::analogRead(__attribute__((unused)) pin_t pin) {
+    // shift registers do not have analog inputs, so always return 0
+    return 0;
+}
+
+template <uint8_t N>
+analog_t ShiftRegisterIn<N>::analogReadBuffered(pin_t pin) {
+    return ShiftRegisterIn<N>::analogRead(pin);
+}
+
+template <uint8_t N>
+void ShiftRegisterIn<N>::begin() {
+    ExtIO::pinMode(clockPin, OUTPUT);
+    ExtIO::pinMode(clockEnablePin, OUTPUT);
+    ExtIO::pinMode(loadPin, OUTPUT);
+
+    ExtIO::digitalWrite(clockPin, LOW);
+    ExtIO::digitalWrite(loadPin, HIGH);
+    ExtIO::digitalWrite(clockEnablePin, HIGH);
+}
+
+template <uint8_t N>
+void ShiftRegisterIn<N>::updateBufferedInputs() {
```

Comment:
From the datasheet:

> For predictable operation the LOW-to-HIGH transition of CE should only take place while CP is HIGH. Also, CP and CE should be LOW before the LOW-to-HIGH transition of PL to prevent shifting the data when PL goes HIGH.

I don't believe this is satisfied here.